src/Raven.Server/Documents/Queries/WhereExpressionVisitor.cs

```diff
@@ -17,69 +19,99 @@ protected WhereExpressionVisitor(string queryText)
+                VisitIn(ie.Source, ie.Values, parameters);
+                return;
+            }
+            if (expression is MethodExpression me)
             {
-                Visit(expression.Left, parameters);
-                Visit(expression.Right, parameters);
+                VisitMethodTokens(me.Name, me.Arguments, parameters);
                 return;
             }
 
-            Debug.Assert(expression.Field != null);
+            if (!(expression is BinaryExpression be))
+            {
+                ThrowUnexpectedExpression(expression);
+                return;// never hit
+            }
 
-            switch (expression.Type)
+            switch (be.Operator)
             {
                 case OperatorType.Equal:
                 case OperatorType.NotEqual:
                 case OperatorType.LessThan:
                 case OperatorType.GreaterThan:
                 case OperatorType.LessThanEqual:
                 case OperatorType.GreaterThanEqual:
-                    VisitFieldToken(QueryExpression.Extract(expression.Field), expression.Value ?? expression.First, parameters);
-                    return;
-                case OperatorType.Between:
-                    VisitFieldTokens(QueryExpression.Extract(expression.Field), expression.First, expression.Second, parameters);
-                    return;
-                case OperatorType.In:
-                case OperatorType.AllIn:
-                    VisitFieldTokens(QueryExpression.Extract(expression.Field), expression.Values, parameters);
-                    return;
-                case OperatorType.Method:
-                    VisitMethodTokens(expression, parameters);
+                    VisitFieldToken(be.Left, be.Right, parameters);
                     return;
+                case OperatorType.And:
+                case OperatorType.AndNot:
+                case OperatorType.Or:
+                case OperatorType.OrNot:
+                    Visit(be.Left, parameters);
+                    Visit(be.Right, parameters);
+                    break;
+
                 default:
-                    throw new ArgumentException(expression.Type.ToString());
+                    ThrowInvalidOperatorType(expression);
+                    break;
             }
         }
 
-        protected ValueTokenType GetValueTokenType(BlittableJsonReaderObject parameters, ValueToken value, bool unwrapArrays)
+        private static void ThrowInvalidOperatorType(QueryExpression expression)
         {
-            var valueType = value.Type;
+            throw new ArgumentException(expression.Type.ToString());
+        }
 
-            if (valueType == ValueTokenType.Parameter)
-            {
-                var parameterName = QueryExpression.Extract(value);
+        private static void ThrowUnexpectedExpression(QueryExpression expression)
+        {
+            throw new InvalidOperationException("Expected binary expression, but got " + expression);
```

Comment:
Maybe better to throw `InvalidQueryException` here?

Reply:
Yes